pkg/ebpf/probes/probes.go

```diff
@@ -81,52 +79,108 @@ func Init(module *bpf.Module) (Probes, error) {
+			UDPSendmsg:                 {event: "udp_sendmsg", probeType: kprobe, fn: "trace_udp_sendmsg"},
+			UDPDisconnect:              {event: "__udp_disconnect", probeType: kprobe, fn: "trace_udp_disconnect"},
+			UDPDestroySock:             {event: "udp_destroy_sock", probeType: kprobe, fn: "trace_udp_destroy_sock"},
+			UDPv6DestroySock:           {event: "udpv6_destroy_sock", probeType: kprobe, fn: "trace_udpv6_destroy_sock"},
+			InetSockSetState:           {event: "sock:inet_sock_set_state", probeType: rawTracepoint, fn: "tracepoint__inet_sock_set_state"},
+			TCPConnect:                 {event: "tcp_connect", probeType: kprobe, fn: "trace_tcp_connect"},
+			ICMPRecv:                   {event: "icmp_rcv", probeType: kprobe, fn: "trace_icmp_rcv"},
+			ICMPSend:                   {event: "__icmp_send", probeType: kprobe, fn: "trace_icmp_send"},
+			ICMPv6Recv:                 {event: "icmpv6_rcv", probeType: kprobe, fn: "trace_icmpv6_rcv"},
+			ICMPv6Send:                 {event: "icmp6_send", probeType: kprobe, fn: "trace_icmp6_send"},
+			Pingv4Sendmsg:              {event: "ping_v4_sendmsg", probeType: kprobe, fn: "trace_ping_v4_sendmsg"},
+			Pingv6Sendmsg:              {event: "ping_v6_sendmsg", probeType: kprobe, fn: "trace_ping_v6_sendmsg"},
 		},
-		module: module,
-	}, nil
+		module:      module,
+		loadNetwork: true,
+	}
+
+	return P, nil
 }
 
-//Attach the probe corresponding to the probe's handle
+// Attach attaches the probe corresponding to the probe's handle.
 func (p *probes) Attach(probe Handle) error {
+
 	if p.probes[probe] == nil {
 		return fmt.Errorf("failed to attach probe: doesn't exist")
 	}
 
 	return p.probes[probe].attach(p.module)
 }
 
-//Detach the probe corresponding to the probe's handle
+// Detach detaches the probe corresponding to the probe's handle.
 func (p *probes) Detach(probe Handle) error {
+
 	if p.probes[probe] == nil {
 		return fmt.Errorf("failed to detach probe: doesn't exist")
 	}
 
 	return p.probes[probe].detach()
 }
 
-// ProbeType is an enum that describes the mechanism used to attach the event
-// Kprobes are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-kp
-// Tracepoints are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#3-tracep
-// Raw tracepoints are explained here: https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#7-raw-tracep
+// DetachAll detaches the probe corresponding to the probe's handle.
+func (p *probes) DetachAll() error {
+
+	for _, probe := range p.probes {
+		fmt.Printf("")
+		err := probe.detach()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DisableNetAutoLoad disables eBPF prog auto load (and attach) for net probes
+func (p *probes) DisableNetAutoload() error {
+	p.loadNetwork = false
+
+	for handle, probe := range p.probes {
+		if int32(handle) >= 1000 { // network probes only
```

Comment:
> We needed to disable network probes because some of the programs will require CAP_NET_ADMIN, and we don't want to force the user to have this capability if he doesn't need to use any network feature of tracee.
> Thinking about it, only the two tc programs we have require this capability, as the others are just kprobes.

But we are loading all eBPF programs by default now, no ? I know we can change that, I was just reflecting what we are doing in a refactor.

We're attaching them based on the dependencies, but we're loading them all. That is why I had the disable autoload for network probes (to accommodate what you said here). To avoid loading the network probes by default.

Reply:
> * I think that instead of looping all the probes, we can just disable the autoload for these two programs.

Ah, ok. I hope CAP_NET_ADMIN is not needed then.

> * If you choose to keep this code, I think it will be more clear to define "1000" as a constant. Also, it is prone to future bugs to have ">= 1000", as in the future someone may add handles starting from, e.g. 2000, and it will be disabled as well.

It sounds good to me if CAP_NET_ADMIN isn't needed. Actually it makes the network probes behave like regular probes, which I'm all in for.

Reply:
Yep, so you're right: Dropping net_admin capability I'm still able to load all probes (but not tcProbes):

If network filters are used:

<img width="1433" alt="image" src="https://user-images.githubusercontent.com/7395852/174156593-3bdbb389-27f7-417f-bd07-2ac54e54f7e2.png">

If not (but all regular network probes are still loaded):

<img width="1432" alt="image" src="https://user-images.githubusercontent.com/7395852/174156531-f6b48b9e-8756-414b-8e8c-57e8d632c416.png">